Clean up OSX input selection

khg8m3r opened the pull request:
Because Quartz handles both keyboard and mouse inputs on OSX, is the OSX keyboard and mouse really needed anymore?
This removes the Keyboard/0 input from Dolphin while still allowing for the OSX controller interface to handle OSXJoystick controllers, I think. I'm still able to see my Logitech F310 in D mode in OSX, but I can't say for sure if other controllers emulate a keyboard.
As for the extra USB inputs, I have no idea how to get rid of those......

MayImilae commented:
This sounds great! macOS has waaaaay too many devices that appear, and half of them don't even work! Here's how it looks on my macbook:
 
![screen shot 2017-08-20 at 11 45 17 pm](https://user-images.githubusercontent.com/6551020/29506640-aac7b620-8601-11e7-9052-8743446133a1.png)

Which ones of those work and which don't? Let's find out! (EDIT: note that the items don't totally match the order and items below, because it friggin changes every single time and whaaatever!)

✖️  Keyboard/0/Apple Internal Keyboard / Trackpad - Only the trackpad works.
❌  Input/0/Apple IR - Doesn't work, obviously, it's the IR remote receiver! Why does this appear here anyway?
❌  Input/0/Apple Internal Keyboard / Trackpad - Doesn't work.
❌  Input/1/Apple Internal Keyboard / Trackpad - Doesn't work either. Not sure what device this is meant to represent... I haven't used an external keyboard in ages.
❌  Input/0/Apple Mikey HID Driver - Doesn't work, obviously, since this is for inputs from headphone/headset buttons! Why is this in the list again?
✖️ Keyboard/0/Keyboard - Well the trackpad works, but the keyboard doesn't, even though it's called "keyboard"? ¯\_(ツ)_/¯
✖️ Keyboard/1/Keyboard - Um, ditto I guess?
✅  Quartz/0/Keyboard & Mouse - Works!

So um, my laptop has all of these options, and _only one of them works the way it should!_ Like, OMG, we need to fix this.

(@khg8m3r sorry for taking over your PR a little bit with a rant!)

ligfx commented:
I've been using the Quartz input since it was put in (by me, hehe) and it works fine. OSXJoystick also works with both of my wired Xbox One Controllers (one appears as an Xbox 360 Controller for some reason, with a different layout, but still works).

The _only_ reason I can see to keep this around would be to support an arbitrary number of keyboards. Maybe someone wants to do local multiplayer with many local keyboards? Idk, seems weird.

Anyways, I'm all for this. Like, I'd vote a **+2** if I could.

khg8m3r commented:
@MayImilae That's exactly the problem! Apple has way to many input devices that aren't actually input devices. I'm only on a Hackintosh so I don't have as many inputs as your does, but it's still to many. It even wants to use my USB headset as an input, even on this PR!
![screen shot 2017-08-21 at 10 40 32 am](https://user-images.githubusercontent.com/31191306/29524214-3bf6e99e-865d-11e7-8970-5d1fe0176b7a.png)

@ligfx Yes, your Quartz implementation was very helpful in pointing me to the files I needed to butcher in order to remove the keyboard :P As for multiple keyboards, I'm not sure how OSX handles it, but at least on Windows they all appear as the same keyboard with no way to tell them apart. I'll grab an extra keyboard today and double-check to see what multiple keyboards look like to Apple in both OSX and Quartz.

ligfx commented:
@dolphin-emu-bot rebuild

MayImilae commented:
>If the Windows backend combines all physical keyboards into a single logical one

Windows takes care of that itself! Even if you have a dozen keyboards plugged in, it will only present one to software, Dolphin included.

khg8m3r commented:
If this can get rebuilt and re-tested, I flipped an if statement in OSX.mm that should get rid of all the excess inputs! 
Plus I think I fixed the clang-format issues.

ligfx commented on Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm:
```diff
@@ -172,21 +167,12 @@ static void DeviceMatchingCallback(void* inContext, IOReturn inResult, void* inS
+    g_controller_interface.AddDevice(std::make_shared<Joystick>(inIOHIDDeviceRef, name));
   }
-#endif
   else
```
Please remove the empty else-block.

lioncash commented:
By the way, you may want to set your commit email to the one you use on github, so you can be more properly verified as the author (this can make it less of a pain if, for whatever reason, we do any licensing related changes way down the line in the distant future and need to contact commit authors).

khg8m3r commented:
@ligfx are you saying that the `kHIDUsage_GD_Joystick` includes the `GamePad` and `MultiAxisController`, or that it needs to be added so it can recognize them?

@lioncash set my email in GitHub preferences, or in my Terminal where I push the commits?

ligfx commented:
It needs to be added so it can recognize them.

(And pretending I'm @lioncash here for a sec: in the Terminal. See https://help.github.com/articles/setting-your-commit-email-address-in-git/ )

khg8m3r commented:
Well I've cleaned up the code and fixed the name, but I have no idea how to properly implement `GamePad` and `MultiAxisController`. IOKit documentation is far above what I can understand in a single night....

ligfx commented:
@khg8m3r Don't worry about implementing anything new, the `Joystick` class should be able to handle most devices. We just need to make sure we allow the devices though: adding those two enums in the same if-statement as the other ones should be enough.

khg8m3r commented:
@ligfx Unfortunately it is not. Adding an if case for `IOHIDDeviceConformsTo(inIOHIDDeviceRef, kHIDPage_GenericDesktop, kHIDUsage_GD_GamePad))` and `MultiaxisController` in OSX.mm throws compile errors. To fix that compile error you need to add a section for `const GamePad* gamepad = dynamic_cast<const GamePad*>(device);`, which throws more compile errors. To fix those errors, I copied Joystick.h and Joystick.mm, changed their name to GamePad, and changed all references from joystick to gamepad, but it still threw compile issues. That was the point at which I went online to try and figure out what IOKit wants and to try and find some examples.

From what I could gleam online, GamePad and MultiAxisController have to be added to Joystick.mm. They turn Joystick, Gamepad, and MAC into an array of some sort and then use that. But translating that to Dolphin is what I'm currently struggling to wrap my head around.

ligfx commented:
@khg8m3r It sounds like you're trying to add new if-cases for each usage page, and new classes. Can you try adding only this?

```diff
--- a/Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm
+++ b/Source/Core/InputCommon/ControllerInterface/OSX/OSX.mm
@@ -167,7 +167,10 @@ static void DeviceMatchingCallback(void* inContext, IOReturn inResult, void* inS
   std::string name = GetDeviceRefName(inIOHIDDeviceRef);
 
   // Add a device if it's of a type we want
-  if (IOHIDDeviceConformsTo(inIOHIDDeviceRef, kHIDPage_GenericDesktop, kHIDUsage_GD_Joystick))
+  if (IOHIDDeviceConformsTo(inIOHIDDeviceRef, kHIDPage_GenericDesktop, kHIDUsage_GD_Joystick) ||
+      IOHIDDeviceConformsTo(inIOHIDDeviceRef, kHIDPage_GenericDesktop, kHIDUsage_GD_GamePad) ||
+      IOHIDDeviceConformsTo(inIOHIDDeviceRef, kHIDPage_GenericDesktop,
+                            kHIDUsage_GD_MultiAxisController))
   {
     g_controller_interface.AddDevice(std::make_shared<Joystick>(inIOHIDDeviceRef, name));
   }
```

That compiles and works fine on my machine.

ligfx commented:
Yep, exactly. You shouldn't need to worry about testing those, since this doesn't change how they're treated: currently, they're handled by the `Joystick` class along with every device that happens to be installed.

khg8m3r commented:
@ligfx Well it compiled fine and still let me use my Logitech F310, so hurray! Plus I installed some xbox 360 drivers and that works as well.

ligfx commented:
@dolphin-emu rebuild

ligfx commented:
@dolphin-emu-bot rebuild
